common/uevent.c

```diff
@@ -248,6 +223,67 @@ uevent_parse(uevent_event_t *uevent)
 	return 0;
 }
 
+uevent_event_t *
+uevent_parse_from_string(const char *uev)
+{
+	IF_NULL_RETVAL_ERROR(uev, NULL);
+
+	int len = strlen(uev);
+
+	uevent_event_t *event = mem_new0(uevent_event_t, 1);
+
+	memcpy(event->msg.raw, uev, len);
+	event->msg_len = len;
+
+	// replace newlines by null bytes
+	for (int i = 0; i < len; i++) {
+		if ('\n' == event->msg.raw[i]) {
+			event->msg.raw[i] = '\0';
+		}
+	}
+
+	uevent_parse(event);
+
+	return event;
+}
+
+static int
+uevent_parse_nl(uevent_event_t *uevent)
+{
+	ASSERT(uevent);
+
+	char *raw_p = uevent->msg.raw;
+
+	DEBUG("raw_p: %p", raw_p);
+	DEBUG("raw_p: %s", raw_p);
+
+	// skip header
+	if (uevent_event_is_udev(uevent)) {
+		/* udev message needs proper version magic */
+		if (uevent->msg.nlh.magic != htonl(UDEV_MONITOR_MAGIC)) {
+			WARN("unrecognized message signature (%x != %x)", uevent->msg.nlh.magic,
+			     htonl(UDEV_MONITOR_MAGIC));
+			return -1;
+		}
+		if (uevent->msg.nlh.properties_off + 32 > uevent->msg_len) {
+			WARN("message smaller than expected (%u > %zd)",
+			     uevent->msg.nlh.properties_off + 32, uevent->msg_len);
+			return -1;
+		}
+		raw_p += uevent->msg.nlh.properties_off;
+	} else if (strchr(raw_p, '@')) {
+		/* kernel message */
+		TRACE("kernel uevent: %s", raw_p ? raw_p : "NULL");
+		raw_p += strlen(raw_p) + 1;
+	} else {
+		/* kernel message */
+		TRACE("no uevent: %s", raw_p);
+		return -1;
+	}
+
+	return uevent_parse(uevent);
```

Comment:
you have to give raw_p to the internal function otherwise, the header skipping code above would
be resettet in uevent_parse()

Reply:
done